### PR TITLE
[fix][security] Add timeout of sync methods and  avoid call sync method for AuthoriationService

### PR DESCRIPTION
**Pull request by codelipenghui**
### Motivation

Add timeout of sync methods for the AuthoriationService to avoid infinite web thread blocking.

image

image

image

### Documentation

Check the box below or label this PR directly.

Need to update docs? 
  
- [x] `no-need-doc` 
(Please explain why)

### CODE REVIEW

**Review by mattisonchao — APPROVED** [review]

**Review by RobertIndie — APPROVED** [review]

**Review by nodece — APPROVED** [review]

**Review by nodece — CHANGES_REQUESTED** [review]
I don't think this PR can fix thread blocking, we should call the async method, not the sync method to authorization.

**Comment by codelipenghui** [comment]
@nodece For the pulsar codebase, we don't have a place to call `allowTenantOperation`, but since they are public methods, which might be called in the plugins? This is my understanding of why the method have have any reference in Pulsar but did not removed yet.

**Comment by codelipenghui** [comment]
@RobertIndie @mattisonchao Please help review again, I have changed the `allowTopicOperation` callers to use the async method `allowTopicOperationAsync` so that we don't have async calls for `AuthoriationService` anymore in the master branch.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 1]
```diff
@@ -3880,46 +3879,53 @@ public static CompletableFuture<PartitionedTopicMetadata> getPartitionedTopicMet
...
-                // throw without wrapping to PulsarClientException that considers: unknown error marked as internal
-                // server error
-                log.warn("Failed to authorize {} on topic {}", clientAppId, topicName, ex);
-                throw ex;
-            }
+        CompletableFuture<Void> authorizationFuture = new CompletableFuture<>();
+        checkAuthorizationAsync(pulsar, topicName, clientAppId, authenticationData)
+                .thenRun(() -> authorizationFuture.complete(null))
+                .exceptionally(e -> {
+                    if (e.getCause() instanceof RestException) {
```
Comment (nodece) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:3886]:
```suggestion
                    if (FutureUtil.unwrapCompletionException(e) instanceof RestException) {
```

Comment (nodece) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:3886]:
```suggestion
                    if (FutureUtil.unwrapCompletionException(e) instanceof RestException) {
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 2]
```diff
@@ -3880,46 +3879,53 @@ public static CompletableFuture<PartitionedTopicMetadata> getPartitionedTopicMet
...
+                        validateAdminAccessForTenantAsync(pulsar,
+                                clientAppId, originalPrincipal, topicName.getTenant(), authenticationData)
+                                .thenRun(() -> {
+                                    authorizationFuture.complete(null);
+                                }).exceptionally(ex -> {
+                                    if (ex.getCause() instanceof RestException) {
```
Comment (nodece) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:3892]:
```suggestion
                                    if (FutureUtil.unwrapCompletionException(ex) instanceof RestException) {
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/lookup/TopicLookupBase.java [hunk 3]
```diff
@@ -220,49 +220,52 @@ public static CompletableFuture<ByteBuf> lookupTopicAsync(PulsarService pulsarSe
...
-                            // request should be redirect to the peer-cluster
-                            if (StringUtils.isBlank(peerClusterData.getBrokerServiceUrl())
-                                    && StringUtils.isBlank(peerClusterData.getBrokerServiceUrlTls())) {
-                                validationFuture.complete(newLookupErrorResponse(ServerError.MetadataError,
-                                        "Redirected cluster's brokerService url is not configured", requestId));
-                                return;
+                checkAuthorizationAsync(pulsarService, topicName, clientAppId, authenticationData).thenRun(() ->
+                                validationFuture.complete(null))
+                        .exceptionally(e -> {
+                            if (e.getCause() instanceof RestException) {
```
Comment (nodece) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/lookup/TopicLookupBase.java:230]:
```suggestion
                            if (FutureUtil.unwrapCompletionException(e) instanceof RestException) {
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/lookup/TopicLookupBase.java [hunk 4]
```diff
@@ -220,49 +220,52 @@ public static CompletableFuture<ByteBuf> lookupTopicAsync(PulsarService pulsarSe
...
+                                log.warn("Failed to authorized {} on cluster {}", clientAppId, topicName);
+                                validationFuture.complete(newLookupErrorResponse(ServerError.AuthorizationError,
+                                        e.getCause().getMessage(), requestId));
```
Comment (nodece) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/lookup/TopicLookupBase.java:233]:
```suggestion
                                        FutureUtil.unwrapCompletionException(e).getMessage(), requestId));
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 5]
```diff
@@ -3880,46 +3879,53 @@ public static CompletableFuture<PartitionedTopicMetadata> getPartitionedTopicMet
...
+                        pulsar.getBrokerService().fetchPartitionedTopicMetadataCheckAllowAutoCreationAsync(topicName))
+                .thenAccept(metadata -> {
+                    if (log.isDebugEnabled()) {
+                        log.debug("[{}] Total number of partitions for topic {} is {}", clientAppId, topicName,
+                                metadata.partitions);
+                    }
+                    metadataFuture.complete(metadata);
+                })
+                .exceptionally(e -> {
+                    metadataFuture.completeExceptionally(e.getCause());
```
Comment (nodece) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:3926]:
```suggestion
                    metadataFuture.completeExceptionally(FutureUtil.unwrapCompletionException(e));
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/lookup/TopicLookupBase.java [hunk 6]
```diff
@@ -220,49 +220,52 @@ public static CompletableFuture<ByteBuf> lookupTopicAsync(PulsarService pulsarSe
...
+                                "Redirected cluster's brokerService url is not configured", requestId));
+                        return;
+                    }
+                    validationFuture.complete(newLookupResponse(peerClusterData.getBrokerServiceUrl(),
+                            peerClusterData.getBrokerServiceUrlTls(), true, LookupType.Redirect, requestId,
+                            false));
+                }).exceptionally(ex -> {
                     validationFuture.complete(
-                            newLookupErrorResponse(ServerError.MetadataError, ex.getMessage(), requestId));
+                            newLookupErrorResponse(ServerError.MetadataError, ex.getCause().getMessage(), requestId));
```
Comment (nodece) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/lookup/TopicLookupBase.java:263]:
```suggestion
                            newLookupErrorResponse(ServerError.MetadataError, FutureUtil.unwrapCompletionException(ex).getMessage(), requestId));
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/lookup/TopicLookupBase.java [hunk 7]
```diff
@@ -220,49 +220,52 @@ public static CompletableFuture<ByteBuf> lookupTopicAsync(PulsarService pulsarSe
...
                     return null;
                 });
             }
         }).exceptionally(ex -> {
-            validationFuture.completeExceptionally(ex);
+            validationFuture.completeExceptionally(ex.getCause());
```
Comment (nodece) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/lookup/TopicLookupBase.java:268]:
```suggestion
            validationFuture.completeExceptionally(FutureUtil.unwrapCompletionException(ex));
```

**Comment by Technoboy-** [comment]
Better to rebase

**Review by Technoboy- — APPROVED** [review]

**Review by nodece — APPROVED** [review]

**Review by RobertIndie — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/lookup/TopicLookupBase.java [hunk 8]
```diff
@@ -220,49 +220,52 @@ public static CompletableFuture<ByteBuf> lookupTopicAsync(PulsarService pulsarSe
...
+                checkAuthorizationAsync(pulsarService, topicName, clientAppId, authenticationData).thenRun(() ->
+                                validationFuture.complete(null))
+                        .exceptionally(e -> {
+                            if (FutureUtil.unwrapCompletionException(e) instanceof RestException) {
+                                log.warn("Failed to authorized {} on cluster {}", clientAppId, topicName);
+                                validationFuture.complete(newLookupErrorResponse(ServerError.AuthorizationError,
+                                        FutureUtil.unwrapCompletionException(e).getMessage(), requestId));
+                            } else {
+                                log.warn("Unknown error while authorizing {} on cluster {}", clientAppId, topicName);
+                                validationFuture.completeExceptionally(e);
```
Comment (Demogorgon314) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/lookup/TopicLookupBase.java:236]:
```suggestion
                            Throwable throwable = FutureUtil.unwrapCompletionException(e);
                            if (throwable instanceof RestException) {
                                log.warn("Failed to authorized {} on cluster {}", clientAppId, topicName);
                                validationFuture.complete(newLookupErrorResponse(ServerError.AuthorizationError,
                                        throwable.getMessage(), requestId));
                            } else {
                                log.warn("Unknown error while authorizing {} on cluster {}", clientAppId, topicName);
                                validationFuture.completeExceptionally(throwable);
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 9]
```diff
@@ -3880,46 +3879,53 @@ public static CompletableFuture<PartitionedTopicMetadata> getPartitionedTopicMet
...
+                        validateAdminAccessForTenantAsync(pulsar,
+                                clientAppId, originalPrincipal, topicName.getTenant(), authenticationData)
+                                .thenRun(() -> {
+                                    authorizationFuture.complete(null);
+                                }).exceptionally(ex -> {
+                                    if (FutureUtil.unwrapCompletionException(ex) instanceof RestException) {
+                                        log.warn("Failed to authorize {} on topic {}", clientAppId, topicName);
+                                        authorizationFuture.completeExceptionally(new PulsarClientException(
+                                                String.format("Authorization failed %s on topic %s with error %s",
+                                                clientAppId, topicName, ex.getCause().getMessage())));
```
Comment (Demogorgon314) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:3896]: Same as the previous comment.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/lookup/TopicLookupBase.java (hunk 8 above)
Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/lookup/TopicLookupBase.java]: Fixed

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java (hunk 9 above)
Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java]: Fixed

**Review by Demogorgon314 — APPROVED** [review]

**Review by eolivelli — APPROVED** [review]
+1

**Comment by BewareMyPower** [comment]
Move the `release/2.8.4` label to https://github.com/apache/pulsar/pull/16831

**Pull request merged**